### ISSUE TITLE
Cannot delete zvol / iscsi datasets - dataset is busy

### ISSUE THREAD

**Issue by jshridha** [democratic-csi/democratic-csi#49]
I'm running on a ubuntu 20.04 ZOL setup and am trying to delete a PVC created using ISCI/ZVOLs. I get the following error message from the csi-driver

```
error: handler error - driver: ControllerZfsGenericDriver method: DeleteVolume error: Error: cannot destroy 'fast/k8s/data/pvc-33cdd659-245a-4d18-aba8-4868e07e9ddc': dataset is busy

{
  code: 13,
  message: "Error: cannot destroy 'fast/k8s/data/pvc-33cdd659-245a-4d18-aba8-4868e07e9ddc': dataset is busy\n"
}
```

Looking in targetcli it appears the iscsi export isn't being deleted:
```
/backstores/block> ls
o- block ...................................................................................................... [Storage Objects: 1]
  o- default-testpvc .............. [/dev/zvol/fast/k8s/data/pvc-33cdd659-245a-4d18-aba8-4868e07e9ddc (5.0GiB) write-thru activated]
    o- alua ....................................................................................................... [ALUA Groups: 1]
      o- default_tg_pt_gp ........................................................................... [ALUA state: Active/optimized
```

Deleting the iscsi target with `delete default-testpvc` seems to clear up the issue. I'm not sure if I have a configuration option set incorrectly in zfs-generic-iscsi.yaml or if something else is going on. Any help would be greatly appreciated, and thanks for the great work. The combination of ZFS+snapshots+kubernetes is gold!

**Comment by jshridha** [comment]
Here is my configuration for reference:

```yaml
csiDriver:
  # should be globally unique for a given cluster
  name: "org.democratic-csi.iscsi"

# add note here about volume expansion requirements
storageClasses:
- name: zfs-generic-iscsi-csi
  defaultClass: false
  reclaimPolicy: Delete
  volumeBindingMode: Immediate
  allowVolumeExpansion: true
  parameters:
    # for block-based storage can be ext3, ext4, xfs
    # for nfs should be nfs
    fsType: ext4
      
    # if true, volumes created from other snapshots will be
    # zfs send/received instead of zfs cloned
    # detachedVolumesFromSnapshots: "false"
    
    # if true, volumes created from other volumes will be
    # zfs send/received instead of zfs cloned
    # detachedVolumesFromVolumes: "false"

  mountOptions: []
  secrets:
    provisioner-secret:
    controller-publish-secret:
    node-stage-secret:
      node-db.node.session.auth.authmethod: CHAP
      node-db.node.session.auth.username: XXXXX
      node-db.node.session.auth.password: XXXXX
    node-publish-secret:
    controller-expand-secret:

# if your cluster supports snapshots you may enable below
volumeSnapshotClasses: []
#- name: zfs-generic-iscsi-csi
#  parameters:
#  # if true, snapshots will be created with zfs send/receive
#  # detachedSnapshots: "false"
#  secrets:
#    snapshotter-secret:

driver:
  config:
    driver: zfs-generic-iscsi
    sshConnection:
      host: XXXXX
      port: 22
      username: csi
      # use either password or key
      # password: ""
      privateKey: |
        XXXXX
    service:
      identity: {}
      controller: {}
      node: {}
    zfs:
      # can be used to override defaults if necessary
      # the example below is useful for TrueNAS 12
      cli:
        sudoEnabled: true
        paths:
          zfs: /usr/sbin/zfs
          zpool: /usr/sbin/zpool
          sudo: /usr/bin/sudo
          chroot: /usr/sbin/chroot
      
      # can be used to set arbitrary values on the dataset/zvol
      # can use handlebars templates with the parameters from the storage class/CO
      datasetProperties:
       "org.freenas:description": "{{ parameters.[csi.storage.k8s.io/pvc/namespace] }}/{{ parameters.[csi.storage.k8s.io/pvc/name] }}"
      #  "org.freenas:test": "{{ parameters.foo }}"
      #  "org.freenas:test2": "some value"

      datasetParentName: fast/k8s/data
      # do NOT make datasetParentName and detachedSnapshotsDatasetParentName overlap
      # they may be siblings, but neither should be nested in the other
      detachedSnapshotsDatasetParentName: fast/k8s/snapshots

      # "" (inherit), lz4, gzip-9, etc
      zvolCompression: ""
      # "" (inherit), on, off, verify
      zvolDedup: ""
      zvolEnableReservation: false
      # 512, 1K, 2K, 4K, 8K, 16K, 64K, 128K default is 16K
      zvolBlocksize: ""

    iscsi:
      shareStrategy: "targetCli"

      # https://kifarunix.com/how-to-install-and-configure-iscsi-storage-server-on-ubuntu-18-04/
      # https://kifarunix.com/how-install-and-configure-iscsi-storage-server-on-centos-7/
      # https://linuxlasse.net/linux/howtos/ISCSI_and_ZFS_ZVOL
      # http://www.linux-iscsi.org/wiki/ISCSI
      # https://bugzilla.redhat.com/show_bug.cgi?id=1659195
      # http://atodorov.org/blog/2015/04/07/how-to-configure-iscsi-target-on-red-hat-enterprise-linux-7/
      shareStrategyTargetCli:
        sudoEnabled: true
        basename: "iqn.2003-01.org.linux-iscsi.ubuntu-19.x8664"
        tpg:
          attributes:
            # set to 1 to enable CHAP
            authentication: 1
            # this is required currently as we do not register all node iqns
            # the effective outcome of this is, allow all iqns to connect
            generate_node_acls: 1
            cache_dynamic_acls: 1
            # if generate_node_acls is 1 then must turn this off as well (assuming you want write ability)
            demo_mode_write_protect: 0
          auth:
            # CHAP
            userid: XXXXX
            password: XXXXX
            # mutual CHAP
            #mutual_userid: "baz"
            #mutual_password: "bar"  
      targetPortal: XXXXXl:3260
      targetPortals: []
      interface: ""

      # MUST ensure uniqueness
      # full iqn limit is 223 bytes, plan accordingly
      # default is "{{ name }}"
      nameTemplate: "{{ parameters.[csi.storage.k8s.io/pvc/namespace] }}-{{ parameters.[csi.storage.k8s.io/pvc/name] }}"
      namePrefix:
      nameSuffix:

```

**Comment by travisghansen** [comment]
Glad that driver is getting a little more use! It should delete the share before attempting to delete the dataset for sure...can you tell from the logs if an attempt is even being made to delete the share or not?

**Comment by travisghansen** [comment]
Actually, I see where the bug is. The `nameTemplate` option was added later and it appears I missed adding that to the delete logic in the generic driver. I'll get a fix into the `next` branch for you to try shortly and you can let me know if it works.

**Comment by travisghansen** [comment]
- https://github.com/democratic-csi/democratic-csi/pull/50/commits/c4979712bd5968ab102d898a484f0ef10f191fc3
- https://github.com/democratic-csi/democratic-csi/runs/1689626029?check_suite_focus=true

When the build finishes go ahead and update the image tag (make sure you have a pull policy set if your nodes already have the `next` image) and let me know how it goes.

**Comment by jshridha** [comment]
I'll give that a shot and let you know. Is there any way to also apply a template to the PVC name rather than generating the random PVC-XXXXX string? I know that's not generally the best design pattern, but it would be useful for my cluster.

**Comment by jshridha** [comment]
I updated the daemonset by `k -n democratic-csi edit deployments.apps zfs-iscsi-democratic-csi-controller` and changed the image to `democraticcsi/democratic-csi:next` after the build was complete

I'm still getting the same result:
```
info: new request - driver: ControllerZfsGenericDriver method: DeleteVolume call: {"_events":{},"_eventsCount":1,"call":{},"cancelled":false,"metadata":{"_internal_repr":{"user-agent":["grpc-go/1.29.1"]},"flags":0},"request":{"secrets":"redacted","volume_id":"pvc-2fc4280e-cd1e-4259-ae85-b5f3b645c8b7"}}
error: handler error - driver: ControllerZfsGenericDriver method: DeleteVolume error: Error: cannot destroy 'fast/k8s/data/pvc-2fc4280e-cd1e-4259-ae85-b5f3b645c8b7': dataset is busy

{
  code: 13,
  message: "Error: cannot destroy 'fast/k8s/data/pvc-2fc4280e-cd1e-4259-ae85-b5f3b645c8b7': dataset is busy\n"
}
```

**Comment by jshridha** [comment]
I reverted the nameTemplate back to default and the issue does in fact go away. It sounds your gut was right and it's related to applying that template.

**Comment by travisghansen** [comment]
OK, I'll have to fire it back up in my dev environment and try it out. The code to delete is this:

```
cd /iscsi
delete ${basename}:${iscsiName}

cd /backstores/block
delete ${iscsiName}
```

Did you happen to notice any output in the logs indicating it tried/failed?

**Comment by jshridha** [comment]
Here's the logs from creation of the PVC to deletion attempt:
```
info: new request - driver: ControllerZfsGenericDriver method: CreateVolume call: {"_events":{},"_eventsCount":1,"call":{},"cancelled":false,"metadata":{"_internal_repr":{"user-agent":["grpc-go/1.29.1"]},"flags":0},"request":{"volume_capabilities":[{"access_mode":{"mode":"SINGLE_NODE_WRITER"},"mount":{"mount_flags":[],"fs_type":"ext4"},"access_type":"mount"}],"parameters":{"fsType":"ext4","csi.storage.k8s.io/pvc/name":"testpvc","csi.storage.k8s.io/pvc/namespace":"default","csi.storage.k8s.io/pv/name":"pvc-ab0c4937-302e-4c04-a604-2a2f1bbc5e79"},"secrets":"redacted","name":"pvc-ab0c4937-302e-4c04-a604-2a2f1bbc5e79","capacity_range":{"required_bytes":"5368709120","limit_bytes":"0"},"volume_content_source":null,"accessibility_requirements":null}}
info: iqn: iqn.2003-01.org.linux-iscsi.ubuntu-19.x8664:default-testpvc
info: new response - driver: ControllerZfsGenericDriver method: CreateVolume response: {"volume":{"volume_id":"pvc-ab0c4937-302e-4c04-a604-2a2f1bbc5e79","capacity_bytes":5368709120,"content_source":null,"volume_context":{"node_attach_driver":"iscsi","portal":"main.local:3260","portals":"","interface":"","iqn":"iqn.2003-01.org.linux-iscsi.ubuntu-19.x8664:default-testpvc","lun":0,"provisioner_driver":"zfs-generic-iscsi"}}}
info: new request - driver: ControllerZfsGenericDriver method: Probe call: {"_events":{},"_eventsCount":1,"call":{},"cancelled":false,"metadata":{"_internal_repr":{"user-agent":["grpc-node/1.24.0-pre1 grpc-c/8.0.0 (linux; chttp2; game)"]},"flags":0},"request":{}}
info: new response - driver: ControllerZfsGenericDriver method: Probe response: {"ready":{"value":true}}
info: new request - driver: ControllerZfsGenericDriver method: DeleteVolume call: {"_events":{},"_eventsCount":1,"call":{},"cancelled":false,"metadata":{"_internal_repr":{"user-agent":["grpc-go/1.29.1"]},"flags":0},"request":{"secrets":"redacted","volume_id":"pvc-ab0c4937-302e-4c04-a604-2a2f1bbc5e79"}}
error: handler error - driver: ControllerZfsGenericDriver method: DeleteVolume error: Error: cannot destroy 'fast/k8s/data/pvc-ab0c4937-302e-4c04-a604-2a2f1bbc5e79': dataset is busy

{
  code: 13,
  message: "Error: cannot destroy 'fast/k8s/data/pvc-ab0c4937-302e-4c04-a604-2a2f1bbc5e79': dataset is busy\n"
}
```

If there's any way to get debug logs showing the commands sent over ssh, let me know and I'll send those too.

**Comment by travisghansen** [comment]
Yeah I'll have to look, it appears I need to add a bit more logging around that as getting it fully logged would be *a lot* of logging currently so I'll just do some testing here and see what I can find. Probably gonna be a day or 2 (headed out on a mini vacation shortly) before I can have a look.

**Comment by jshridha** [comment]
No problem. Thanks for taking a look. At the end of the day, the ISCSI names don't matter to me much so I'm going to revert to the default template. Just throwing this out there - perhaps removing that feature would be an option?

**Comment by travisghansen** [comment]
I could potentially remove it yeah, but I'm guessing it's a super-simple fix and shouldn't take much to fix up.

**Comment by travisghansen** [comment]
OK, fixed up in `next` and images are building:

- https://github.com/democratic-csi/democratic-csi/runs/1699652273?check_suite_focus=true

It will only work with newly-created volumes so a fresh start is recommended :( try it out and let me know if it works properly for you.

**Comment by jshridha** [comment]
That worked great! Thanks for the fix.

**Comment by travisghansen** [comment]
Landed in `latest` and `v1.0.1`
